### PR TITLE
When the space is not enough to contain the pop up use the best fit 

### PR DESCRIPTION
**Pull request by ChrisCinelli**
I was having problems on mobile  ((#58) where the pop up was always going out of viewport.

This PR improves the positioning in case there are no zones where the pop up can completely fit.
The orientation is chosen based on the zone that cut the smallest area of the pop up (see an example here: http://jsfiddle.net/uthyZ/ )

`preferPlace` is also kept in consideration. For example if `preferPlace='column'`, the pop up will be positioned either above or below according to the zone that show the largest part of the popup.

There are also:
- a commit that remove the dependency on the assign polyfill that is automatically managed by babel.
- a commit for onOuterAction so it can receive the `mousedown` or `touchstart` event that generated it.
- a commit that fix problems with building because `babel-preset-react` was not a dependency

### CODE REVIEW

**Comment by jasonkuhrt** [comment]
@ChrisCinelli Thanks for putting this together. I am going to try and review early in the new year.

File: lib/index.js [hunk 1]
```diff
@@ -396,7 +396,7 @@ const Popover = createClass({
       !this.containerEl.contains(event.target) &&
       !this.targetEl.contains(event.target)
     )
-    if (isOuterAction) this.props.onOuterAction()
+    if (isOuterAction) this.props.onOuterAction(event)
```
Comment (jasonkuhrt) [inline lib/index.js]: Is this change needed?

File: lib/layout.js [hunk 2]
```diff
@@ -135,7 +135,7 @@ const getPreferenceType = (preference) => (
 /* Dimension Fit Checks */
 
 const fitWithinChecker = (dimension) => (domainSize, itemSize) => (
-  domainSize[dimension] > itemSize[dimension]
+  domainSize[dimension] >= itemSize[dimension]
```
Comment (jasonkuhrt) [inline lib/layout.js]: I think this makes sense but is it needed for this PR?

File: lib/layout.js [hunk 3]
```diff
@@ -174,11 +172,21 @@ const pickZone = (opts, frameBounds, targetBounds, size) => {
...
     { side: `start`, standing: `left`, flow: `row`, order: -1, w: t.x, h: f.y2 },
   ]
-
+  
+  /* Order the zones by the amount of popup that would be cut out if that zone is used.
+     The first one in the array is the one that cuts the least amount.
+      
+     const area = size.w * size.h  // Popup area is constant and it does not change the order
+  */
+  zones.forEach((z) => {
```
Comment (jasonkuhrt) [inline lib/layout.js]: This ordering is only used if there are no available zones. Therefore we should only compute it when there are no available zones.

**Comment by jasonkuhrt** [comment]
This looks almost ready to merge. Let me know about the few questions asked.

File: package.json [hunk 4]
```diff
@@ -14,8 +14,7 @@
   "dependencies": {
     "css-vendor": "^0.3.1",
     "debug": "^2.1.3",
-    "lodash.throttle": "^3.0.3",
-    "object.assign": "^4.0.1"
```
Comment (jasonkuhrt) [inline package.json]: Is this necessary for this PR?

**Comment by jasonkuhrt** [comment]
This looks almost ready to merge. Let me know about the few questions asked.

File: lib/index.js (hunk 1 above)
Comment (ChrisCinelli) [inline lib/index.js]: I was hiding the pop up on click outside the modal. I noticed that the pop up was closing to easily on tap on mobile. I was not able to scroll up or down if necessary.
So I wanted to close only when the mouse was used. This change let me use:

    _handleOuterAction = (e) => {
      if (e.type == 'mousedown')
        this._handleOnClose();
    }   

Why not put it in? It is backward compatible.

File: lib/layout.js (hunk 2 above)
Comment (ChrisCinelli) [inline lib/layout.js]: Well, when the dimension of the pop up are 100%, the popover does not fit, while instead it should  fit.
Is it a problem that I fix other stuff in the same PR?

File: lib/layout.js (hunk 3 above)
Comment (ChrisCinelli) [inline lib/layout.js]: As I wrote in the comment here: https://github.com/littlebits/react-popover/pull/60/files#diff-77ab7858fd1708d83f99faf433c11a56R213 this is actually always useful because ordering the zones by best fit. You have a few `myZones[0]` that are taking the first zone available. If we always reorder the zones by best fit, it will always pick the best fit. So for example when `opt.preferPlace="column"`, the selected zone will be the best fit between `above` and `below`. 

Considering that the `h` and `w` are already available and you have 4 zones, we are talking of a computation that is taking less than 1ms. 

If we are trying to optimize, we can actually remove the `availZones` computation, since the will be the first ones after sorting.

File: package.json (hunk 4 above)
Comment (ChrisCinelli) [inline package.json]: This will make the code better. Babel has its own implementation of `Object.assign` that is shared in the whole compiled code. Adding the module `object.assign` is going to add multiple copies of the code that does the polyfill of the `Object.assign`.

Again, is it a problem having improvements in a PR even if not relative to the PR title? 
If you look at the single commits in this PR they already properly separated this and the `this.props.onOuterAction(event)` above.

File: lib/index.js (hunk 1 above)
Comment (jasonkuhrt) [inline lib/index.js]: :+1: Thanks! Nitpick on the next PR you make (hopefully there is a next one! : ) try to keep this good but unrelated improvements out of the PR. It makes reviewing proposals easier for me and it also makes changelog links more readable.

File: lib/layout.js (hunk 2 above)
Comment (jasonkuhrt) [inline lib/layout.js]: Ditto comment above. Thanks!

File: lib/layout.js (hunk 3 above)
Comment (jasonkuhrt) [inline lib/layout.js]: Thanks, makes sense as a lightweight improvement. To add to this discussion, I don't think I would consider this a solution to #14 which I think deserves more robust dedicated consideration. I know that was not your intent here but I think its worth referencing that old issue. I don't know what a more robust algorithm would be exactly but I can imagine that for instance it would know to treat a absolutely smaller area that is more balanced on both dimensions as preferable. This warrants a diagram to make clear but hopefully the hint has been successfully suggested. Anyways, again, thanks.

Comment (jasonkuhrt) [inline lib/layout.js]: Resolved.

File: lib/layout.js (hunk 2 above)
Comment (jasonkuhrt) [inline lib/layout.js]: Resolved.

File: lib/index.js (hunk 1 above)
Comment (jasonkuhrt) [inline lib/index.js]: Resolved.

File: package.json (hunk 4 above)
Comment (jasonkuhrt) [inline package.json]: > Babel has its own implementation of Object.assign that is shared in the whole compiled code.

Hm, can you clarify please?

React Popover is compiled to ES5 before publishing so what users consume via NPM is ES5 code. When I use React Popover in e.g. a Babel-based JS2015 app that is _again_ compiled to ES5 before deployment **I fail to understand how Object.assign polyfill code is not simply duplicated**. Contrast with NPM which knows how to keep a flat dependency tree as of version 3.

Comment (jasonkuhrt) [inline package.json]: Another solution that I have so far opted against is to let the user supply their own polyfill but I think its less friendly for the user. Either they have to read the README which states this or we use NPM Peer Dependencies which still suffers from forcing us to choose which library to use.

Comment (ChrisCinelli) [inline package.json]: I did not consider the intermediate compilation. Hopefully, webpack 2 and Babel will be able to use modules written in ES6 without [hack](https://github.com/kriasoft/react-starter-kit/blob/master/tools/webpack.config.js#L162). Ok I will remove it from the pull request.

**Comment by ChrisCinelli** [comment]
Ok, I removed the commit with the `assign` change.

File: package.json (hunk 4 above)
Comment (jasonkuhrt) [inline package.json]: Thanks.

:+1: Resolved.

**Comment by jasonkuhrt** [comment]
@ChrisCinelli Thank-you. I will cut a patch release today.

**Comment by jasonkuhrt** [comment]
https://github.com/littlebits/react-popover/releases/tag/v0.4.4

File: lib/layout.js [hunk 5]
```diff
@@ -190,17 +198,27 @@ const pickZone = (opts, frameBounds, targetBounds, size) => {
...
       zone[preferenceType] === opts.preferPlace
     ))
     if (preferredAvailZones.length) return preferredAvailZones[0]
+    
+    // If there are not areas where the pop up fit completely, it uses the prefered ones
+    // in order from the one the fit better
+    const preferredZones = zones.filter((zone) => (
+      zone[preferenceType] === opts.preferPlace
+    ))
+    if (preferredZones.length) return preferredZones[0]
```
Comment (stephen) [inline lib/layout.js]: Hm, this seems like a regression, @jasonkuhrt, @ChrisCinelli?

This will pick from `preferredZones` ignoring whether or not there are other available zones (which is a case handled by the return statement after this). To match what the comment says, it seems like this check should be
```diff
- if (preferredZones.length) return preferredZones[0]  
+ if (!availZones.length && preferredZones.length) return preferredZones[0]  
```

Here is a gif of the behavior I'm seeing (with `preferPlace` as `below`), which is rectified by the above diff:
falling-off-issue
